File: Trees/102-levelOrderTraversal.java
// problem link: https://leetcode.com/problems/binary-tree-level-order-traversal/

package Trees;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

class Solution { // 1ms beats 89%
    public List<List<Integer>> levelOrder(TreeNode root) {

        List<List<Integer>> result = new ArrayList<>();

        if (root == null) {
            return result;
        }

        Queue<TreeNode> queue = new ArrayDeque<TreeNode>();

        queue.offer(root);

        while (!queue.isEmpty()) {

            List<Integer> currentLevel = new ArrayList<>();

            int levelSize = queue.size();

            for (int i = 0; i < levelSize; i++) {

                TreeNode currentNode = queue.poll();

                if (currentNode.left != null) {
                    queue.offer(currentNode.left);
                }
                if (currentNode.right != null) {
                    queue.offer(currentNode.right);
                }

                currentLevel.add(currentNode.val);
            }

            result.add(currentLevel);

        }
        return result;
    }
}

File: Trees/102-levelOrderTraversal.cpp
#include <vector>

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};
class DFSPreOrderTraversal {  // 0ms beats 100%
  std::vector<std::vector<int>> levelOrder(TreeNode *root) {
        std::vector<std::vector<int>> result;
    if (root == nullptr) {
      return result;
    }

    levelOrder(root, 0, result);
    return result;
  }

  void levelOrder(TreeNode *currentNode, int height,
                  std::vector<std::vector<int>> &result) {
    if (currentNode == nullptr) {
      return;
    }

    if (height == result.size()) {
      result.emplace_back();
    }
    result.at(height).push_back(currentNode->val);

    levelOrder(currentNode->left, height + 1, result);
    levelOrder(currentNode->right, height + 1, result);
  }
};